Meteorite World Generation Expansion for Dimensions and Loot Tables

Gamingb3ast opened the pull request:
In line with a previous pull request I made, this pull request focuses on the Meteorites.

-----------
# Changes:

Created a config option to add Valid/Invalid Blocks to the list of possible blocks that the meteorites can spawn on
Added a comment to the config to make the process as straightforward as possible.
Includes many error messages to help config editors debug.
## Meaning users can now make meteorites spawn in more biomes and dimensions (or fewer).


Expanded the dimension whitelist config to allow overriding debris (helps it fit into the world even more)
This works with block metadata!
Also included a DEFAULT tag, which falls back onto the old system for that specific block, so if you wish to remove gravel, you can do so easily.
Added a note to the config to make the process as straightforward as possible.
## Meaning users can now modify exactly how the meteorites will look in any dimensions (Screenshots included)


Modified registration.java to work with these changes, added a new method to get a block from the GameRegistry with a String input.
Modified MeteoriteBlockPutter.java and Fallout.java to correspond to these changes and generate blocks depending on the config.

-----------

# Screenshots

Mars
![2025-04-18_14 06 11](https://github.com/user-attachments/assets/a999a9aa-be3d-4900-a76b-e99e01e1f9dc)

![2025-04-18_14 06 20](https://github.com/user-attachments/assets/978f48ac-d138-4893-ac02-1ee0c167de42)
-----------------------
Moon
![2025-04-18_14 04 39](https://github.com/user-attachments/assets/41dc0b45-e7ba-45dd-b24b-25497640e266)

![2025-04-18_14 05 15](https://github.com/user-attachments/assets/3a370af7-046f-4d85-a097-6c3ea49bee3e)


More is possible, but this is just what I made with this simple config


```cfg
worldgen {
    D:meteoriteClusterChance=1
    S:meteoriteDimensionWhitelist <
        0, DEFAULT, DEFAULT, DEFAULT, DEFAULT, DEFAULT
        -28, GalacticraftCore:tile.moonBlock:4, GalacticraftCore:tile.moonBlock:4, GalacticraftCore:tile.moonBlock:3, GalacticraftCore:tile.moonBlock:3, GalacticraftCore:tile.moonBlock:5
        -29, GalacticraftMars:tile.mars:9, GalacticraftMars:tile.mars:4, GalacticraftMars:tile.mars:6, minecraft:sand:1, GalacticraftMars:tile.mars:5
     >
    S:meteoriteInvalidSpawnBlocks <
        examplemod:example_block
     >
    D:meteoriteSpawnChance=1
    S:meteoriteValidSpawnBlocks <
        examplemod:example_block
        GalacticraftCore:tile.moonBlock
        GalacticraftMars:tile.mars
     >
    I:minMeteoriteDistance=50
    I:quartzOresClusterAmount=15
    I:quartzOresPerCluster=4
    D:spawnChargedChance=0.07999998331069946
}
```

Dream-Master commented:
@Gamingb3ast I like the idea but maybe we need to re consider the loot in the meteor chests can have some more advanced items. Only on planets not on the overworld

Gamingb3ast commented:
> @Gamingb3ast I like the idea but maybe we need to re consider the loot in the meteor chests can have some more advanced items. Only on planets not on the overworld

Hey there, glad to hear back, I'm not fully sure what you mean, are you suggesting the ability to modify or disable loot chests on a dimensional basis? I may be misunderstanding you.

Dream-Master commented:
> > @Gamingb3ast I like the idea but maybe we need to re consider the loot in the meteor chests can have some more advanced items. Only on planets not on the overworld
> 
> Hey there, glad to hear back. I'm not fully sure what you mean. Are you suggesting the ability to modify or disable loot chests on a dimensional basis? I may be misunderstanding you.

What I mean is that you should get different loot on different planets, not just the forms. That would make the search for meteorites on planets more meaningful.

Gamingb3ast commented:
> > > @Gamingb3ast I like the idea but maybe we need to re consider the loot in the meteor chests can have some more advanced items. Only on planets not on the overworld
> > 
> > Hey there, glad to hear back. I'm not fully sure what you mean. Are you suggesting the ability to modify or disable loot chests on a dimensional basis? I may be misunderstanding you.
> 
> What I mean is that you should get different loot on different planets, not just the forms. That would make the search for meteorites on planets more meaningful.

I absolutely agree! I'll go ahead and implement this into the pull request as soon as I can! I'll likely put the loot tables into a separate config for versatility and keeping the main config less crowded.

Dream-Master commented:
@Gamingb3ast let me know when this can be tested and be done

Gamingb3ast commented:
> @Gamingb3ast let me know when this can be tested and be done

Will do, I've gotten the loot JSON to work, I just need to make a default config and do some personal testing before handing it to ya'll.

I'll let you know soon if I don't rewrite it another 3 times.

Gamingb3ast commented:
@Dream-Master 

I have finalized it all. 
It works on Mars, the default configs work fine, but they won't generate the exact same loot tables as before. I think it's more interesting, but feel free to change it to suit the mod's needs. I've also added documentation to the JSON, but I can make things clearer for anyone editing the defaults if necessary. 
Meteorites generate as normal, and all seems well in my testing for the last few hours.

### The one placeholder left...
...is the glowstone dust item as a fallback item in the AEJSONEntry.getItemStack method. It can stay, or it can be changed. I used something that makes it clear when the loot fails to generate. It is currently used as the placeholder for both failure in finding an item in the ore dictionary (which occurs often with default configs if only using AE2) and general failure. This can be changed if you see it as unfit to have glowstone in the chest loot.

I would also advise changing or removing the loot tables for Mars if there is concern that it is too OP. I simply used something that would make it clear how the configs work and could be used.


# Changes
-------------

- No more manual deserialization on JSON creation.
   - Switched to the instance variable for certain methods.
- Moved to a 2d arraylist object-based Entry Map:
  -  Allows for multiple unique loot tables in one dimension.(is essentially the opposite of an exclusive group)
  -  Reworked all methods to support the 2d array.
  -  Removed unnecessary methods.
  -  Loot tables are selected based on total entry weight.
- Cleaned up AEJSONEntry.java
  -  Added another constructor
  - Added itemstack accessor method
- Fixed file creation on initialization event
- Completed chest loot creation in MeteoritePlacer.java
   - Removed old loot placement 
   - Added general functionality as expected from the JSON.
- Reworked defaults
- Added OreDictionary Support
- Made glowstone the base item stack for when an entry fails to load one
- Made dimensions that are not a part of the config use the currently set overworld config rather than the default one.
- Finalized error logging
- Optimized Loot Addition
- Cleaned up error logs
- Added documentation for the JSON
- Allowed parameters to be optional
- Fixed non-inclusive stack size selection

Dream-Master commented:
@Gamingb3ast Ok cool will try it later. So I guess a config file will be generated which I can edit ? Or is it a config file per dim ?

Gamingb3ast commented:
> @Gamingb3ast Ok cool will try it later. So I guess a config file will be generated which I can edit ? Or is it a config file per dim ?

It will generate a single config file when starting a world or reloading the game. The single file will have all dimensions. The default config has 2 dimensions by default (Earth and Mars) it's a little cramped but that's simply how the JSON generates, it can be spread out and made easier to read without breaking functionality. There's a comment at the beginning with just some basic instructions on how it works.

Dream-Master commented:
> There is one more thing I've thought of this morning that I'd like to add to this PR. The ability to change spawn chance and distance for meteorites on a per dimension basis so that one may be able to have finer control over meteorites. I've also thought of this as someone may make a really good loot table for a certain dimension but want meteorites to be rarer than on another dimension where they're more common and vice versa.
> 
> The reason I want to implement it with this PR is that I feel everything that was done here is for the one goal of adding meteorites more easily to other dimensions, which is a short and simple thing to put on a changelog.
> 
> I'll implement this soon as I've checked the code and it shouldn't take too long to do.

yea good idea

Gamingb3ast commented:
> yea good idea

All done. It's gotten quite versatile. Worked fine as far as I tested it on the Moon and Mars. It should be my last addition to World Gen for now.

serenibyss commented on src/main/java/appeng/core/AEConfig.java:
```diff
@@ -165,20 +165,28 @@ public AEConfig(final File configFile) {
+                .getStringList();
         this.meteoriteSpawnChance = this.get("worldGen", "meteoriteSpawnChance", this.meteoriteSpawnChance)
-                .getDouble(this.meteoriteSpawnChance);
+                .getStringList();
```
Have changes like this been tested? How does the forge config system handle the type here being changed while leaving the config name the same?

Gamingb3ast replied:
I'm assuming you mean if you keep the old config while updating the mod? It will end up crashing with the custom exception:
java.lang.RuntimeException: MISSING VALUE meteoriteSpawnChance FOR OVERWORLD
Its the intended behavior.
Since the spawn chances and distances need to be linked to a dimension now, I had to switch it over to a string so the config could be something similar to this:

```cfg
S:minMeteoriteDistance <
        0=707
        -28=20
        -29=100
     >
```
Here, the distances for the overworld, moon, and Mars, respectively, will be 707, 20, and 100.
If the configs are changed to the new format by the user or simply deleted and then regenerate,d then it will function just fine until the user makes an error while entering a value. For example, not including a value for the overworld (Which is used as the default) or doing ```cfg -28= ```

serenibyss commented on src/main/java/appeng/core/AEJSONConfig.java:
```diff
@@ -0,0 +1,269 @@
+            "    \"-29\": [   this is the start of a list of loot tables for another dimension, and so on", "       [",
+            "         {", " }" };
+    @SerializedName("dimension_loot_tables")
+    private Map<String, ArrayList<ArrayList<AEJSONEntry>>> dimension_loot_tables = new HashMap<>();
```
Could you name this something more typical, like `dimensionLootTables`? The `@SerializedName` annotation should keep it from causing any problems in the json representation

serenibyss commented on src/main/java/appeng/core/AEJSONConfig.java:
```diff
@@ -0,0 +1,269 @@
+        try {
+            FileUtils.writeStringToFile(file, GSON.toJson(instance), Charset.defaultCharset());
+        } catch (Exception e) {
+            System.err.println(
```
Could this use the AELog?

serenibyss commented on src/main/java/appeng/core/AEJSONConfig.java:
```diff
@@ -0,0 +1,269 @@
+            AEJSONConfig.instance = loaded;
+            return loaded;
+        } catch (Exception e) {
+            System.err.println(
```
This as well

serenibyss commented on src/main/java/appeng/core/AEJSONConfig.java:
```diff
@@ -0,0 +1,269 @@
+            } else if (dimension_loot_tables.containsKey("0")) {
+                return dimension_loot_tables.get("0");
+            } else {
+                System.err.println(
```
Here as well

serenibyss commented on src/main/java/appeng/core/AEJSONConfig.java:
```diff
@@ -0,0 +1,269 @@
+                return createDefaultConfig().getTablesForDimension(0);
+            }
+        } else {
+            System.err.println(
```
Here as well

serenibyss commented on src/main/java/appeng/core/AEJSONConfig.java:
```diff
@@ -0,0 +1,269 @@
+    public ArrayList<AEJSONEntry> getWeightedLootTable(int dimID, Random rand) {
+        ArrayList<ArrayList<AEJSONEntry>> loot_tables = instance.getTablesForDimension(dimID);
+        if (loot_tables == null || loot_tables.isEmpty()) {
+            System.err.println(
```
Here as well

serenibyss commented on src/main/java/appeng/core/AEJSONConfig.java:
```diff
@@ -0,0 +1,269 @@
+                return loot_tables.get(i);
+            }
+        }
+        System.err.println(
```
Here as well

serenibyss commented on src/main/java/appeng/core/AEJSONEntry.java:
```diff
@@ -0,0 +1,94 @@
+    public AEJSONEntry(String item, Integer meta_data, Integer min_value, Integer max_value, Integer weight,
+            Integer exclusiveGroupID) {
+        if (item == null) {
+            System.err.println("AE2: itemName is required! | Error: While loading JSON");
```
I see there are a lot of these, I won't comment on the rest

Gamingb3ast replied:
Not sure what you're referring to here...

serenibyss replied:
Just the usage of `System.err` or `System.out` rather than using `AELog`

Gamingb3ast replied:
Oh I see now. Yes I will use that.

serenibyss commented on src/main/java/appeng/core/AEJSONEntry.java:
```diff
@@ -0,0 +1,94 @@
+    /**
+     * Primary constructor with optional parameters.
+     *
+     * @param item             Required. Format: "modid:item" or "ore:OredictName"
```
On the extremely small chance that some mod somewhere uses `ore` as a mod id, could this be formatted slightly differently? Maybe `modid:item / ore#OredictName`, up to your discretion

Gamingb3ast replied:
Sure, that slight off chance is something that should be accounted for. I'll set that up now.

serenibyss commented on src/main/java/appeng/core/Registration.java:
```diff
@@ -847,8 +848,51 @@ void postInit(final FMLPostInitializationEvent event) {
+                        "AE2: Error while whitelisting dimension from string: " + dimension
+                                + " | Error: Too Many or Few Parameters!");
+            } else {
+                int dimID = -32727; // -1*Short.MAX_VALUE-50, this value represents failure to find the dimension in
```
Could you make this value a static final constant?

Gamingb3ast replied:
Will do, likely won't cause issues.

serenibyss commented on src/main/java/appeng/core/Registration.java:
```diff
@@ -861,4 +905,34 @@ void postInit(final FMLPostInitializationEvent event) {
+    public static Block getBlockFromString(String entry) {
+        if (entry == null) {
+            System.err.println("AE2: Error getting block from string | Error: String is null!");
+            return net.minecraft.init.Blocks.air; // Air represents failure to find block
```
Could you import `Blocks` rather than using the fully qualified reference?

Gamingb3ast replied:
Oh, didn't notice that, will do.

Gamingb3ast replied:
Blocks seems to conflict with ```import appeng.api.definitions.Blocks;``` so I'll just import ```import static net.minecraft.init.Blocks.air;``` as that is the only constant I use in blocks.

serenibyss commented on src/main/java/appeng/worldgen/MeteoritePlacer.java:
```diff
@@ -296,59 +321,81 @@ private void placeMeteorite(final IMeteoriteWorld w, final int x, final int y, f
+                    InventoryAdaptor ap = InventoryAdaptor.getAdaptor(te, ForgeDirection.UP);
+
+                    int dimID = w.getWorld().provider.dimensionId;
+                    ArrayList<AEJSONEntry> loot_table = new ArrayList<>(
+                            AEJSONConfig.instance.getWeightedLootTable(dimID, lootRng));
+                    Map<Integer, ArrayList<AEJSONEntry>> exlcusion_table_map = new HashMap<>();
```
Could you rename these variables as well?

Gamingb3ast replied:
Which variables are you referring to, specifically?

serenibyss replied:
`loot_table` and `exlcusion_table_map` (which I see has a typo in it now), would be more typical to name them `lootTable` and `exclusionTableMap`

Gamingb3ast replied:
I see, will do

serenibyss commented on src/main/java/appeng/worldgen/MeteoriteWorldGen.java:
```diff
@@ -33,14 +33,50 @@ public void generate(final Random rng, final int chunkX, final int chunkZ, final
         if (WorldGenRegistry.INSTANCE.isWorldGenEnabled(WorldGenType.Meteorites, world)) {
             // Find the meteorite grid cell corresponding to this chunk
-            final int gridCellSize = Math.max(8, AEConfig.instance.minMeteoriteDistance);
+            int defaultGridCellSize = -3124;
+            int gridCellSize = -3124;
```
What is the significance of `-3124`? Could you make it a static final constant and put a comment on it saying what it represents and why its this number?

Gamingb3ast replied:
Will do, just chose a random number which would be unlikely to be selected and wasn't too large.

serenibyss commented on src/main/java/appeng/core/AEJSONConfig.java:
```diff
@@ -0,0 +1,269 @@
+public class AEJSONConfig {
+
+    public static AEJSONConfig instance;
+    /**
+     * Parameters: item: TYPE: STRING DEFAULT VALUE: NONE. PARAMETER IS REQUIRED DESCRIPTION: The name of the item as
+     * used by the /give command EXAMPLE: "minecraft:dirt" meta_data: TYPE: NON-NEGATIVE INTEGER DEFAULT VALUE: 0
+     * DESCRIPTION: Metadata value for the item, the 3rd parameter in the /give command EXAMPLE: 3 min_value: TYPE:
+     * NON-NEGATIVE INTEGER DEFAULT VALUE: 0 DESCRIPTION: The minimum amount of that item that is inserted when that
+     * item is selected (SETTING TO 1 DOES NOT GUARANTEE THAT ITEM IN EVERY CHEST) EXAMPLE: 4 max_value: TYPE:
+     * NON-NEGATIVE INTEGER DEFAULT VALUE: 1 DESCRIPTION: The maximum amount of that item that is inserted when that
+     * item is selected EXAMPLE: 72 weight: TYPE: NON-NEGATIVE INTEGER DEFAULT VALUE: 1 DESCRIPTION: The weighted chance
+     * the item is chosen. (Weights are compared between entries with the same exclusive group) EXAMPLE: 40
+     * exclusiveGroupID: TYPE: INTEGER DEFAULT VALUE: -1 DESCRIPTION: exclusiveGroupID lets you group entries so only
+     * one from the group can appear, chosen based on weight. If multiple entries share the same group ID, only one will
+     * be picked. If an entry is the only one in its group, it's guaranteed to appear. Setting this parameter to -1 will
+     * cause it to ignore exclusivity EXAMPLE: 154
+     *
+     * Notes: The Key Value: For the dimension_loot_tables a key value is required, this value represents the
+     * dimensionID for the loot tables. For example, -29 is Galacticraft Mars, the loot tables under that id will only
+     * spawn in that dimension (if meteorites spawn there) The list of loot tables: The chance of a loot table within a
+     * dimension being selected is weighted by the total weight of every entry in that loot table against every other
+     * loot tables' total weights Format: { "dimension_loot_tables": { "0": [ This is the start of a list of loot tables
+     * [ this is the start of a loot table { ENTRY }, { ANOTHER ENTRY } ], this is the end of a loot table [ this is the
+     * start of another separate loot table { ENTRY }, { ANOTHER ENTRY } ] this is the end of a loot table ], this is
+     * the end of the list of loot tables "-29": [ this is the start of a list of loot tables for another dimension, and
+     * so on [ { }
+     */
```
If you surround this with a spotless disable block, like:
```java
// spotless:off
// ... your code comment ...
// spotless:on
```
you can format this more nicely. Though up to you if you want to do this

Gamingb3ast replied:
I didn't know that was available, I will use that!

serenibyss reviewed (commented):
Thanks for the great work! I have some preliminary review comments

Gamingb3ast commented:
> Thanks for the great work! I have some preliminary review comments
Alright, I think that's everything you mentioned in the review.

Gamingb3ast commented:
Just checking up on this PR :)

lordIcocain commented:
@Gamingb3ast 
![image](https://github.com/user-attachments/assets/c59bd81a-8e74-4f1d-8e87-b85858e3e6cd)

Dream-Master commented:
@Gamingb3ast btw did it need a config change and can you make some examples. it seems also that Ae2 Quartz are generated in overworld. We just disabled it.

Gamingb3ast commented:
> @Gamingb3ast btw did it need a config change and can you make some examples. it seems also that Ae2 Quartz are generated in overworld. We just disabled it.

Sure, I just had it generate a default config that was similar to the old loot chests but also a bit more unique and showed all the possibilities with this new config, I'll put an example here and just put lots of comments on each line to make it as clear as possible exact how it works and what can be done.

Gamingb3ast commented:
> @Gamingb3ast 
> ![image](https://github.com/user-attachments/assets/c59bd81a-8e74-4f1d-8e87-b85858e3e6cd)

Not really sure what this is...
As far as I'm aware this base fork of AE2 doesn't have a file with all meteorite spawn locations, maybe im wrong, the default config has them generate pretty infrequently 700 blocks apart, I didn't change that. Could more context be provided if this is ae2 from my PR? Feel free to contact me on discord just to keep this PR less cluttered. My user is just @Gamingb3ast

Gamingb3ast commented:
> I just tested in last nightly and seems like meteors not generating at all or meteor compass broken.

I am installing the latest nightly right now, I will test it and get back with a proper config but from one glance it looks like it uses the old format for world gen configs. A simple config refresh should fix it but to make it a bit easier on ya'll I will give you a copy paste for the world gen section so as to remove the need to write the entire config again.

Dream-Master commented:
ok thanks will test it

Gamingb3ast commented:
I forgot to mention that you can also do per dimension spawn distance and chances by doing something like this:

```conf
S:minMeteoriteDistance <
        0=707
        29=707
     >
```

Gamingb3ast commented on src/main/java/appeng/core/Registration.java:
```diff
@@ -850,8 +855,51 @@ void postInit(final FMLPostInitializationEvent event) {
+                        blockList[i - 1] = Registration.getBlockFromString(entry[i]);
+                        try {
+                            String[] split = entry[i].split(":");
+                            metaList[i - 1] = (split.length == 2 ? 0 : Integer.parseInt(split[2]));
```
No, `split` would be formatted as either minecraft:grass or minecraft:grass:4 as an example.
If `split` ends up being 2 (aka only one `:`) then there is no metadata specified and 0 is assumed. Anything else would parse and use the integer located after the second `:`.

Gamingb3ast commented:
> the game crashes when scrolling down in AE2 config in game

Hmmm, does it crash in a specific category, since I couldn't replicate the bug.
![image](https://github.com/user-attachments/assets/7f334481-6acd-484c-92a9-e7d5275530e1)
![image](https://github.com/user-attachments/assets/add72ac8-7fc0-478e-9490-ef47621564fc)

NeOzay commented:
> > the game crashes when scrolling down in AE2 config in game
> 
> Hmmm, does it crash in a specific category, since I couldn't replicate the bug. ![image](https://private-user-images.githubusercontent.com/86311666/460321574-7f334481-6acd-484c-92a9-e7d5275530e1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTEyMjc0MTQsIm5iZiI6MTc1MTIyNzExNCwicGF0aCI6Ii84NjMxMTY2Ni80NjAzMjE1NzQtN2YzMzQ0ODEtNmFjZC00ODRjLTkyYTktZTdkNTI3NTUzMGUxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjI5VDE5NTgzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0MmIwM2U5N2VkN2E5NGI3YTBjMGIxYzNlMDg0NDU3MDIwZmM3MDA1NTc0NzA1NzJkNmIyMWQyYzQyYWJhMTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vnnDuxhZgNPVWvlkgBntwwHMio8LCINwqS__tAAMHzo) ![image](https://private-user-images.githubusercontent.com/86311666/460321599-add72ac8-7fc0-478e-9490-ef47621564fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTEyMjc0MTQsIm5iZiI6MTc1MTIyNzExNCwicGF0aCI6Ii84NjMxMTY2Ni80NjAzMjE1OTktYWRkNzJhYzgtN2ZjMC00NzhlLTk0OTAtZWY0NzYyMTU2NGZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjI5VDE5NTgzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ0YzczZWZhYzkwMWQ2NmFmYmU5ZTMzN2UyOTBhNjNhNzEzODFkNzNhMDZiMDQ1OTA5YzY4NDQ4MGY2MjIwNGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.LLWi_ROu8WbBM3rK6vBFMjNwtgvT6zVCOntOtoH_uys)

mmh, i do (in dev env)
check if in last commit -> title screen -> Mods -> select AE2 -> select config -> scroll down
and
```
java.lang.NumberFormatException: For input string: "0, DEFAULT, DEFAULT, DEFAULT, DEFAULT, DEFAULT"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
	at java.base/java.lang.Integer.parseInt(Integer.java:662)
	at java.base/java.lang.Integer.valueOf(Integer.java:989)
Caused by: java.lang.NumberFormatException: For input string: "0, DEFAULT, DEFAULT, DEFAULT, DEFAULT, DEFAULT"

	at Launch//net.minecraftforge.common.config.ConfigElement.getDefaults(ConfigElement.java:242)
	at Launch//cpw.mods.fml.client.config.GuiConfigEntries$ArrayEntry.isDefault(GuiConfigEntries.java:751)
	at Launch//cpw.mods.fml.client.config.GuiConfigEntries.areAllEntriesDefault(GuiConfigEntries.java:302)
	at Launch//cpw.mods.fml.client.config.GuiConfigEntries$CategoryEntry.isDefault(GuiConfigEntries.java:1415)
	at Launch//cpw.mods.fml.client.config.GuiConfigEntries$ListEntryBase.drawEntry(GuiConfigEntries.java:1599)
	at Launch//cpw.mods.fml.client.config.GuiConfigEntries$CategoryEntry.drawEntry(GuiConfigEntries.java:1372)
	at Launch//net.minecraft.client.gui.GuiListExtended.drawSlot(GuiListExtended.java:35)
	at Launch//net.minecraft.client.gui.GuiSlot.drawSelectionBox(GuiSlot.java:481)
	at Launch//net.minecraft.client.gui.GuiSlot.drawScreen(GuiSlot.java:348)
	at Launch//cpw.mods.fml.client.config.GuiConfig.drawScreen(GuiConfig.java:344)
	at Launch//net.minecraft.client.renderer.EntityRenderer.updateCameraAndRender(EntityRenderer.java:1137)
	at Launch//net.minecraft.client.Minecraft.runGameLoop(Minecraft.java:1067)
	at Launch//net.minecraft.client.Minecraft.run(Minecraft.java:962)
	at Launch//net.minecraft.client.main.Main.main(Main.java:164)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at System//net.minecraft.launchwrapper.Launch.rfb$realLaunch(Launch.java:250)
	at System//net.minecraft.launchwrapper.Launch.launch(Launch.java:35)
	at System//net.minecraft.launchwrapper.Launch.main(Launch.java:60)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.gtnewhorizons.retrofuturabootstrap.Main.main(Main.java:207)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at System//net.minecraftforge.gradle.GradleStartCommon.launch(GradleStartCommon.java:99)
	at System//GradleStart.main(GradleStart.java:40)
```

Gamingb3ast commented:
Ok so I've merged all the changes from the current master branch of this repo and tested it again. I really don't know what to say it seems to work fine going through all the categories in the config. I'm sorry I couldn't replicate the bug. My assumption is that it had something to do with an existing, old config file that would cause the parsing error.

It's unlikely since I imagine the testing was done in a clean build and all the default values are correctly parsed, but it does seem like the config expected an integer even though it's specified to be a string list. I don't imagine any case where that would happen besides an old existing config where it looked like this:

```cfg

I:meteoriteDimensionWhitelist <
        0
     >
```

It might have imposed the default value but not changed the I type and caused something like this:

```cfg
I:meteoriteDimensionWhitelist <
        0, DEFAULT, DEFAULT, DEFAULT, DEFAULT, DEFAULT
     >
```

rather than what it should be which is this:

```cfg
S:meteoriteDimensionWhitelist <
        0, DEFAULT, DEFAULT, DEFAULT, DEFAULT, DEFAULT
     >
```

NeOzay commented:
works correctly after cleaning the config

Gamingb3ast commented:
@Dream-Master 

Yup will do, I just want to take some time today to look into a nicer way than just renaming the config values, I won't look too hard so I'll be ready soon. Let me know on discord (@Gamingb3ast) if you'll want any help with the configs for the modpack, I tired to make the instructions as clear as possible but not sure how well it worked out.

Also let me know if there's anything else at all!

NeOzay commented:
@Gamingb3ast if you haven't started to solve it, i will solve the problem today

Gamingb3ast commented:
@NeOzay 
I spent about 30 minutes tinkering with different ways to get the current type for a config key and found that it's not easy and definitely not worth it. 

Changing the names is definitely the best way to go but it doesn't resolve the issue of the game not letting the ingame configs  be opened.

I'm currently working on a way to remove the old values and I think I've found it, will push another commit in 20 minutes or so

Gamingb3ast reviewed (commented):
Works 👍

NeOzay reviewed (approved):
🎉🎉🎉🎉
